# Decision Trees on Iris Dataset

### Submitted by - 
> Name : Shumbul Arifa \
> Roll No.: 181CO152 \
> Date : 15-01-2021


**Problem Statement:** 
#### To create a decision tree, using the below criterions for splitting the attributes:
1. Quinlan's ID3  (using information gain) 
2. C4.5 (using gain ratio) 
3. CART (using gini index)

# ID3

## 1) Install required libraries

In [1]:
!pip install graphviz
from sklearn.datasets import load_iris
from sklearn import tree
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import graphviz 
import pandas as pd

## 2) Load IRIS dataset

In [2]:
iris = load_iris()

## 3) Split data & fit the model

In [3]:
X, y = load_iris(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 25, random_state = 10)
clf = tree.DecisionTreeClassifier(criterion='entropy')
clf = clf.fit(X_train, y_train)
clf = clf.fit(X_train, y_train)

## 4) Accuracy

In [4]:
print("Accuracy of ID3 Model: ",  clf.score(X_test, y_test))

Accuracy of ID3 Model:  0.96


# C4.5


## 1) Install required libraries

In [5]:
!pip install chefboost


## 2) Load the data

In [6]:
data = load_iris()
df = pd.DataFrame(data.data, columns=data.feature_names)
df['Decision'] = pd.Series(data.target)
df['Decision'] = df['Decision'].replace({0:"type0" , 1: "type1",2:"type2"});
train=df.sample(frac=0.75,random_state=10)
test=df.drop(train.index)

## 3) Build the model

In [7]:
from chefboost import Chefboost as chef
config = {'algorithm': 'C4.5'}
model = chef.fit(train, config ,test)

C4.5  tree is going to be built...
-------------------------
finished in  0.5835182666778564  seconds
-------------------------
Evaluate  train set
-------------------------
Accuracy:  77.67857142857143 % on  112  instances
Labels:  ['type1' 'type2' 'type0']
Confusion matrix:  [[18, 5, 0], [20, 32, 0], [0, 0, 37]]
Decision  type1  => Accuray:  77.6786 %, Precision:  78.2609 %, Recall:  47.3684 %, F1:  59.0164 %
Decision  type2  => Accuray:  77.6786 %, Precision:  61.5385 %, Recall:  86.4865 %, F1:  71.9101 %
Decision  type0  => Accuray:  100.0 %, Precision:  100.0 %, Recall:  100.0 %, F1:  100.0 %
-------------------------
Evaluate  validation set
-------------------------
Accuracy:  81.57894736842105 % on  38  instances
Labels:  ['type0' 'type1' 'type2']
Confusion matrix:  [[13, 0, 0], [0, 6, 1], [0, 6, 12]]
Decision  type0  => Accuray:  100.0 %, Precision:  100.0 %, Recall:  100.0 %, F1:  100.0 %
Decision  type1  => Accuray:  81.5789 %, Precision:  85.7143 %, Recall:  50.0 %, F1:  63

## 4) Analysis

The accuracy of C4.5 is 78.3% on training dataset

# Gini - index

## 1) Install Required Libraries

In [8]:
from sklearn.datasets import load_iris
from sklearn import tree
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import graphviz 
import pandas as pd

## 2) Load Iris dataset

In [9]:
iris = load_iris()
X, y = load_iris(return_X_y=True)

## 3) Split data 

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 25, random_state = 10)
#25% data for testing

## 4) Fit the model

In [11]:
clf = tree.DecisionTreeClassifier(criterion='gini')
clf = clf.fit(X_train, y_train)

## 5) Accuracy 

In [12]:
print(clf.score(X_test, y_test))

0.96


# Overall Analysis




Observed accuracy of each algorithm is as follows:
ID3 - 96%
C4.5 - 78.3%
CART - 96%

Thus, ID3 and CART show accuracy >90%, so these two are best suited for the decision tree algorithm on the iris dataset. Whereas, C4.5 is the least accurate, thus not suitable for the same as compared to two other algorithms.

**Therefore, ID3 and CART are the best performing Decision Tree Algorithms on the Iris dataset**